File: leetcode/17. Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if len(digits) == 0:
            return []
        dic = {'2':'abc', '3':'def', '4':'ghi', '5':'jkl', '6':'mno', '7':'pqrs', '8':'tuv', '9':'wxyz'}
        res = ['']
        for i, digit in enumerate(digits):
            tmp = []
            for c in dic[digit]:
                for element in res:
                    tmp.append(element + c)
            res = tmp
        return res




class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if digits == '':
            return []
        dic = {'2':'abc',
                '3': 'def',
                '4': 'ghi',
                '5': 'jkl',
                '6': 'mno',
                '7': 'pqrs',
                '8': 'tuv',
                '9': 'wxyz'}
        res = []
        digits = list(digits)
        
        def helper(idx, curr):
            if idx == len(digits):
                res.append(curr)
                return
            
            for c in dic[digits[idx]]:
                helper(idx+1, curr+c)
        
        helper(0, '')
        return res

File: leetcode/17. Letter Combinations of a Phone Number.kt
val keyboard: Map<Char, String> = mapOf(
    '2' to "abc",
    '3' to "def",
    '4' to "ghi",
    '5' to "jkl",
    '6' to "mno",
    '7' to "pqrs",
    '8' to "tuv",
    '9' to "wxyz")

class Solution {
    fun letterCombinations(digits: String): List<String> {
        val res = mutableListOf<String>()
        if (digits.length == 0) return res
        fun getCombinations(idx: Int, curr: MutableList<Char>) {
            if (idx == digits.length) {
                res.add(curr.joinToString(""))
                return
            }
            for (c in keyboard[digits[idx]]!!) {
                curr.add(c)
                getCombinations(idx+1, curr)
                curr.removeLast()
            }
        } 
        getCombinations(0, mutableListOf<Char>())
        return res
    }
}
